scipy/interpolate/src/__fitpack.h

```diff
@@ -1,64 +1,178 @@
+#pragma once
+#include <iostream>
+#include <tuple>
+#include <vector>
+#include "../_build_utils/src/fortran_defs.h"
+
+#define DLARTG F_FUNC(dlartg, DLARTG)
+
+
+namespace fitpack {
+
+
 /*
- * B-spline evaluation routine.
+ * 1D and 2D array wrappers, with and without boundschecking
  */
 
-static inline void
-_deBoor_D(const double *t, double x, int k, int ell, int m, double *result) {
-    /*
-     * On completion the result array stores
-     * the k+1 non-zero values of beta^(m)_i,k(x):  for i=ell, ell-1, ell-2, ell-k.
-     * Where t[ell] <= x < t[ell+1].
-     */
-    /*
-     * Implements a recursive algorithm similar to the original algorithm of
-     * deBoor.
-     */
-    double *hh = result + k + 1;
-    double *h = result;
-    double xb, xa, w;
-    int ind, j, n;
-
-    /*
-     * Perform k-m "standard" deBoor iterations
-     * so that h contains the k+1 non-zero values of beta_{ell,k-m}(x)
-     * needed to calculate the remaining derivatives.
-     */
-    result[0] = 1.0;
-    for (j = 1; j <= k - m; j++) {
-        memcpy(hh, h, j*sizeof(double));
-        h[0] = 0.0;
-        for (n = 1; n <= j; n++) {
-            ind = ell + n;
-            xb = t[ind];
-            xa = t[ind - j];
-            if (xb == xa) {
-                h[n] = 0.0;
-                continue;
-            }
-            w = hh[n - 1]/(xb - xa);
-            h[n - 1] += w*(xb - x);
-            h[n] = w*(x - xa);
-        }
+
+// Bounds checking
+template<bool boundscheck> inline void _bcheck(ssize_t index, ssize_t size, ssize_t dim);
+
+
+template<>
+inline void _bcheck<true>(ssize_t index, ssize_t size, ssize_t dim) {
+    if (!((0 <= index) && (index < size))){
+        auto mesg = "Out of bounds with index = " + std::to_string(index) + " of size = ";
+        mesg = mesg + std::to_string(size) + " in dimension = " + std::to_string(dim);
+        throw(std::runtime_error(mesg) );
     }
+}
+
+template<>
+inline void _bcheck<false>(ssize_t index, ssize_t size, ssize_t dim) { /* noop*/ }
+
+
+// Arrays: C contiguous only
+template<typename T, bool boundscheck=true>
+struct Array1D
+{
+    T* data;
+    ssize_t nelem;
+    T& operator()(const ssize_t i) {
```

Comment:
While it'd be nice to avoid `operator()` indeed, I thought it'd be nice to be consistent for `Array1D` and `Array2D`? 
OTOH, this whole homemade stuff should probably be replaced with what https://github.com/scipy/scipy/pull/20320 implements. I'll ask over there.

Reply:
Let's focus on the ``std::mdspan`` implementation from here. Do you know if that works well with NumPy arrays?